tcp_endpoint/src/tcp_endpoint/RosTCPClientThread.py

```diff
@@ -130,14 +133,27 @@ def run(self):
             print("No data for a message size of {}, breaking!".format(full_message_size))
             return
 
-        if destination not in self.source_destination_dict.keys():
-            error_msg = "Topic/Service destination '{}' does not exist in source destination dictionary {} "\
-                .format(destination, self.source_destination_dict.keys())
+        if destination.startswith('__'):
+            if destination not in self.tcp_server.special_destination_dict.keys():
+                error_msg = "System message '{}' is undefined! Known system calls are: {}"\
+                    .format(destination, self.tcp_server.special_destination_dict.keys())
+                self.conn.close()
+                self.tcp_server.send_unity_error(error_msg)
+                raise TopicOrServiceNameDoesNotExistError(error_msg)
+            else:
+                ros_communicator = self.tcp_server.special_destination_dict[destination]
+                # this feels pretty hacky...
```

Comment:
Would be useful to call out _what_ feels hacky about this, what a potential fix would look like, and why it needs to be done this way for now.

Reply:
Oh yeah, I wrote that so we'd talk about it in this review :-)
I haven't written much Python before, and I was struggling to figure out the "right" way to implement the HandshakeService. My original idea was to have all the special_destination_dict components work just like regular ROS services... but for the HandshakeService to do its job it needs to know the IP of the incoming connection. No normal service needs that. Is there an elegant way for it to get access to that but still provide the same interface as other services?

Reply:
Hmmm, on the one hand it's useful to be able to re-use the serialization/deserialization code that the ROS message generation provides, but other than that, does the handshake service even need to behave like ROS service? It's kind of hijacking the ROS message pipeline to pass messages that aren't really intended to be exposed to ROS. We've got a websocket connection established between the TCP-Connector and TCP-Endpoint and the ability to send arbitrary json blobs back and forth... would it be difficult to set this up more like a traditional web service using json for a pseudo-rpc setup?

Reply:
Uh, well, we're not sending json, our connection is sending "messages", which are binary blobs in the serialization format that ROS uses. What you're suggesting sounds like a pretty major rewrite.

Reply:
Right but what I'm saying is that these messages don't necessarily need to be serialized by ROS convention; they _could_ be arbitrary json. They don't even really need to come in on the same port as the ROS messages (and maybe it'd be better if they didn't?)  Anyway I think you're right that this would mandate quite a lot of new code so it seems out of scope for this change. I _am_ a little concerned that this sets a precedent for implementing this kind of messaging in a potentially hacky way, but maybe we can mitigate that by calling it out and linking in a comment to a ticket for implementing something more specifically tailored for this class of messages?